lib/toy/object.rb

```diff
@@ -1,4 +1,5 @@
 require "toy/behaviours"
+require "toy/object_instance"
```

Comment:
Similarly, we should strictly `require "toy/class"` here because `Object.class` returns `Class`.

…But I think it’s inevitable we’re going to have a problem with circular requires, because, per your [cool diagram](https://user-images.githubusercontent.com/22918438/139278815-1d457843-dd9e-4b55-aef4-4039104954df.jpg):

* `Class.superclass` returns `Module`
* `Module.superclass` returns `Object`
* `Object.class` returns `Class`

So I don’t see how it’s possible for us to ever load `class.rb`, `module.rb` and `object.rb` in any linear order if they each strictly require their dependencies, because the dependency relationships really do form a cycle.

Therefore we probably need to make peace with either defining all three constants in the same file so they can all refer to each other without needing to involve `require`, or else be less strict about needing to `require` every constant that’s referenced in a file and rely on some other convention (e.g. “you must `require "toy"` if you want to use any of the `Toy::` classes”) to make sure everything gets loaded. WDYT?

Reply:
Yeah, this is a good question huh? 😁  I'm leaning towards the second option -- "be less strict about needing to require every constant that’s referenced in a file and rely on some other convention". Perhaps it makes sense to autoload all of the constants in the `Toy` namespace instead of requiring at the top of every file?

ie. We could define `lib/toy.rb` to be:
```ruby
module Toy
  autoload(:Behaviours, './toy/behaviours.rb')
  autoload(:ClassInstance, './toy/class_instance.rb')
  autoload(:Class, './toy/class.rb')
  autoload(:ModuleInstance, './toy/module_instance.rb')
  autoload(:Module, './toy/module.rb')
  autoload(:ObjectInstance, './toy/object_instance.rb')
  autoload(:Object, './toy/object.rb')
end
```

I'm not sure what the convention is on where `lib/toy.rb` would be required though 🤔

Reply:
This is a good idea! Maybe we could `require 'toy'` from any file under `lib/toy` which depends on any other? Then any user (e.g. a test) could just `require 'toy/module'` (or whatever) as normal, and get any other constants autoloaded if/when they’re needed. I think. I haven’t tried it.

Reply:
Cool! I'll give it a go and see what I can come up with 🙇‍♀️

Reply:
Okay, so I actually ended up going a different direction here. If you hate it, let me know 😆 

Instead of doing `require 'toy'` in all of our different files, I decided to _just_ require it in the `test_helper`, since that's already consistently required by our  "users" (just tests in this case). I imagine if we had a console binstub we could also require it there, etc... Then the code files / tests don't need to require anything, because the test helper autoloads all of the relevant constants. 

I pushed a couple more commits sticking with this approach. Let me know what you think!

Reply:
That sounds very sensible!

The tradeoff is that if a (hypothetical) new user wants to use, say, `Toy::Module`, they can’t just `require 'toy/module'` — they have to somehow know that `require 'toy'` is the right thing to do. But that's not inherently a big deal, and e.g. Rack [makes this same choice](https://github.com/rack/rack#label-Usage):

> …you should not need require paths under `rack`. If you require paths under `rack` without requiring `rack` itself, things may not work correctly.

So you’re in good company at least.

Reply:
> Rack makes this same choice

Cool! Yeah, I was pretty sure I'd seen this pattern elsewhere which nudged me in this direction. I realized shortly after writing this that we did the same thing for Packwerk! (A [`lib/packwerk` file](https://github.com/Shopify/packwerk/blob/main/lib/packwerk.rb) that gets required in the [`test_helper`](https://github.com/Shopify/packwerk/blob/main/test/test_helper.rb#L9))